### PR TITLE
Feature/59180 rescheduling on changing week days and holidays 

### PR DESCRIPTION
**Pull request by toy**
# Ticket

[OP#63901](https://community.openproject.org/wp/63901) (part of [OP#59180](https://community.openproject.org/wp/59180), follows #18523 and #18824)

# What are you trying to accomplish?
Automatic rescheduling of phases when working days and holidays are change in admin interface

## Screenshots

# What approach did you choose and why?

# Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

### CODE REVIEW

File: app/models/project/phases/scopes/covering_dates_and_days_of_week.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  class_methods do
+    # Fetches all phases that cover specific days of the week, and/or specific dates.
+    #
+    # The period considered is from the phase start date to the due date.
+    #
+    # @param dates Date[] An array of the Date objects.
+    # @param days_of_week number[] An array of the ISO days of the week to
+    #   consider. 1 is Monday, 7 is Sunday.
+    def covering_dates_and_days_of_week(days_of_week: [], dates: [])
```
Comment (ulferts) [inline app/models/project/phases/scopes/covering_dates_and_days_of_week.rb]: This method seems to return all the phases which:
* span a certain weekday OR
* span a certain date

So I think the method should rather be named `covering_dates_or_days_of_week`

File: app/models/work_packages/scopes/covering_dates_and_days_of_week.rb [hunk 2]
```diff
@@ -59,30 +59,25 @@ def where_covers_periods(work_packages_periods_cte, days_of_week, dates)
...
+            SELECT 1
+            FROM UNNEST(ARRAY[:days_of_week]::INT[]) AS target_dow
+            WHERE (target_dow + 7 - EXTRACT(ISODOW FROM start_date)::INT) % 7 <= (end_date - start_date)
+          )
+          OR
+          -- Check if the range covers any of the provided dates
+          EXISTS (
+            SELECT 1
+            FROM unnest(Array[:dates]::DATE[]) AS target_date
+            WHERE target_date BETWEEN start_date AND end_date
```
Comment (ulferts) [inline app/models/work_packages/scopes/covering_dates_and_days_of_week.rb:82]: 👍 for cleaning this up in lockstep with the sibling scope.

File: app/workers/apply_working_days_change_job_base.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    User.execute_as user do
+      apply_working_days_change
+    end
+  end
+
+  private
+
+  def apply_working_days_change
+    fail NoMethodError, "Must be overridden in subclass"
```
Comment (ulferts) [inline app/workers/apply_working_days_change_job_base.rb:56]: Not overly important but why is a NoMethodError used here instead of a NotImplementedError?

File: app/workers/projects/phases/apply_working_days_change_job.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Projects::Phases::ApplyWorkingDaysChangeJob < ApplyWorkingDaysChangeJobBase
+  private
+
+  def apply_working_days_change
+    user = User.current
```
Comment (ulferts) [inline app/workers/projects/phases/apply_working_days_change_job.rb]: I think this is here uselessly.

File: app/workers/projects/phases/apply_working_days_change_job.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    end
+  end
+
+  def applicable_phases
+    days_of_week = changed_days.keys
+    dates = changed_non_working_dates.keys
+
+    Project::Phase
+      .active # TODO: should visible be used?
+      .covering_dates_and_days_of_week(days_of_week:, dates:)
```
Comment (ulferts) [inline app/workers/projects/phases/apply_working_days_change_job.rb]: Since a user calling this job is an admin, using the `visible` scope would only change the scope to only include active projects. Archived projects would be included as permissions are not active in there, even for admins. 

That might be a worthwhile approach, though, since it would then not change the scheduling of already completed projects. However, if we follow along this path, we would have to reschedule the phases when reactivating a project. That is a use case, that while being a valid one, I currently judge to be of lesser importance so I would currently not support it. I also wouldn't know where to make the cut. Because one could also argue that phases that are scheduled in the past should not be rescheduled. To me however, that argument only is worth to follow up on when we support validity periods for the non working days settings.

TL;DR: Using the `active` scope seems fine to me.

File: app/workers/projects/phases/apply_working_days_change_job.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class Projects::Phases::ApplyWorkingDaysChangeJob < ApplyWorkingDaysChangeJobBase
+  private
+
+  def apply_working_days_change
+    user = User.current
+
+    applicable_phases.distinct.pluck(:project_id).each do |project_id|
```
Comment (ulferts) [inline app/workers/projects/phases/apply_working_days_change_job.rb]: The `distinct` IMO is unnecessary since the scope should not return duplicates. On the other hand, adding a `.uniq` after the `pluck(:project_id)` would save unnecessary iterations. 

Here, however, it might be better to use something like 

```
Project.where(id: applicable_phases.select(:project_id)).each do |project|
```

That way, the check for the existence of the project can be be omitted as well.

File: app/workers/projects/phases/apply_working_days_change_job.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      project = Project.find_by(id: project_id)
+      next unless project
+
+      phases = project.available_phases
+      next if phases.empty?
+
+      from = phases.first.start_date
```
Comment (ulferts) [inline app/workers/projects/phases/apply_working_days_change_job.rb]: The first phase might have its start date not set but other phases in the same project can still have dates. Because of that, I think it safer to use 

```
from = phases.pluck(:start_date, :finish_date).flatten.compact.min
```

File: app/services/project_life_cycle_steps/reschedule_service.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    def persist(service_call)
+      reschedule_phases(**params)
+
+      service_call
+    end
+
+    def reschedule_phases(phases:, from:)
+      phases.each do |phase|
+        next unless phase.date_range_set?
+        next unless phase.duration&.positive? # not updated using service that sets duration
```
Comment (ulferts) [inline app/services/project_life_cycle_steps/reschedule_service.rb:52]: Currently the caller is responsible to filter out phases that should not be rescheduled. On the other hand, some selection is taking place in here. Why is `active` not filtered for here?

**Review by ulferts — CHANGES_REQUESTED** [review]
The implementation looks solid and well structured. 

I did not manage to run through a complete set of test cases yet but found one edge case that is worth to take a look at.

### Having project phases with a phase in the middle not having a date set

image

When adding a non working day within the time frame spanned by the other phases, the rescheduling will break.

File: app/workers/apply_working_days_change_job_base.rb (hunk 3 above)
Comment (toy) [reply app/workers/apply_working_days_change_job_base.rb:56]: It is incorrect to use `NotImplementedError` in such case, at it is «Raised when a feature is not implemented on the current platform». Defining method and raising same `NoMethodError` error as when it would not be defined is for documenting and giving more info in the raised exception. So it would be better to use that or introduce an `AbstractMethodError` and use it in all other places, but given that this looks like the only place where `NoMethodError` is used instead of `NotImplementedError`, it can be changed later across the codebase if other developers agree

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 4 above)
Comment (toy) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: It is needed for `RescheduleService`

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 6 above)
Comment (toy) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: `applicable_phases` returns a scope for phases, so can return multiple phases for same project and `distinct` applies to the select from `pluck`. Compare SQLs for simpler queries without and with distinct:
```ruby
Project::Phase.pluck(:project_id)
# SELECT "project_phases"."project_id" FROM "project_phases"
```
```ruby
Project::Phase.distinct.pluck(:project_id)
# SELECT DISTINCT "project_phases"."project_id" FROM "project_phases"
```

I went with iterating ids of projects and loading projects later to not load all projects in memory and because slight decrease of performance loading projects one by one should not be very important here. Alternative is probably to use `Project.where(id: applicable_phases.select(:project_id)).find_each`

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 7 above)
Comment (toy) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: Good point. First I though that `covering_dates_and_days_of_week` should not allow that, but it returns also phases with only one date set if it is on the affected day of week or date. Still it is better to force querying `phases`, as they are needed afterwards anyway

File: app/services/project_life_cycle_steps/reschedule_service.rb (hunk 8 above)
Comment (toy) [reply app/services/project_life_cycle_steps/reschedule_service.rb:52]: Good point, while now I rely on input to provide only active phases, it is better to add the check to be less confusing in the future. Also it is a reminder that phases are also not checked for order, so if a scope is used that would return them not in definition order, they will be rescheduled in incorrect order, but it felt odd to load definitions here only to ensure order.

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 7 above)
Comment (toy) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: I'm also not sure if it makes sense to consider `finish_date` if `start_date` is not set

**Comment by toy** [comment]
@ulferts Thank you for thorough review, I either applied changes or replied, so please have a look

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 4 above)
Comment (ulferts) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: True, I didn't see the connection. I guess it was moved outside of the loop for performance reasons. However, I think the overhead of calling User.current on a call to 

```
ProjectLifeCycleSteps::RescheduleService.new(user: User.current, project:)
```

would be minimal and would increase readability.

File: app/workers/apply_working_days_change_job_base.rb (hunk 3 above)
Comment (ulferts) [reply app/workers/apply_working_days_change_job_base.rb:56]: Thanks for the clarification. I wasn't aware of that definition so I misused the `NotImplementedError` in the past it seems.

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 6 above)
Comment (ulferts) [reply app/workers/projects/phases/apply_working_days_change_job.rb]:
```
Project.where(id: applicable_phases.select(:project_id)).find_each
```

Is what I would work with, then. To me, the extra loading and checking of the project produces noise.

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 7 above)
Comment (ulferts) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: When not considering the finish_date, something like this can happen (edge case admittedly):

Have a set of phases where one only has its finish date set. 

image

Add a non working days to the finish date set for the phase. 

Now add a start date to the phase which only had a finish date:

image

I am not sure, if this should be prevented by the automatic rescheduling, though. It could set the date to be on the next non working day. But since it requires the user to interact with the phase again to raise the error it might be acceptable to have the invalid date kept. What do you think, @toy ?

**Review by ulferts — APPROVED** [review]
Good work @toy. From my point of view, this is good to merge. 

There is this one edge case still open which I am undecided about but this can either decided to be ignored or be fixed later on. 

So once the CI is green, go ahead and merge 🎉

File: app/workers/projects/phases/apply_working_days_change_job.rb (hunk 7 above)
Comment (toy) [reply app/workers/projects/phases/apply_working_days_change_job.rb]: I would keep it as is for now, I think updating finish date when phase has no start date may be more unexpected than allowing it to become invalid
